core/src/server/subscription.rs

```diff
@@ -0,0 +1,475 @@
+	/// - `Err(err)` if the connection or subscription was closed.
+	///
+	/// # Cancel safety
+	///
+	/// This method is cancel-safe and dropping a future loses its spot in the waiting queue.
+	pub async fn send(&self, msg: SubscriptionMessage) -> Result<(), DisconnectError> {
+		// Only possible to trigger when the connection is dropped.
+		if self.is_closed() {
+			return Err(DisconnectError(msg));
+		}
+
+		let json = sub_message_to_json(msg, SubNotifResultOrError::Result, &self.uniq_sub.sub_id, self.method);
+		self.inner.send(json).await.map_err(Into::into)
+	}
+
+	/// Similar to to `SubscriptionSink::send` but only waits for a limited time.
+	pub async fn send_timeout(&self, msg: SubscriptionMessage, timeout: Duration) -> Result<(), SendTimeoutError> {
+		// Only possible to trigger when the connection is dropped.
+		if self.is_closed() {
+			return Err(SendTimeoutError::Closed(msg));
+		}
+
+		let json = sub_message_to_json(msg, SubNotifResultOrError::Result, &self.uniq_sub.sub_id, self.method);
+		self.inner.send_timeout(json, timeout).await.map_err(Into::into)
+	}
+
+	/// Attempts to immediately send out the message as JSON string to the subscribers but fails if the
+	/// channel is full or the connection/subscription is closed
+	///
+	///
+	/// This differs from [`SubscriptionSink::send`] where it will until there is capacity
+	/// in the channel.
+	pub fn try_send(&mut self, msg: SubscriptionMessage) -> Result<(), TrySendError> {
+		// Only possible to trigger when the connection is dropped.
+		if self.is_closed() {
+			return Err(TrySendError::Closed(msg));
+		}
+
+		let json = sub_message_to_json(msg, SubNotifResultOrError::Result, &self.uniq_sub.sub_id, self.method);
+		self.inner.try_send(json).map_err(Into::into)
+	}
+
+	/// Returns whether the subscription is closed.
+	pub fn is_closed(&self) -> bool {
+		self.inner.is_closed() || !self.is_active_subscription()
+	}
+
+	/// Completes when the subscription has been closed.
+	pub async fn closed(&self) {
+		// Both are cancel-safe thus ok to use select here.
+		tokio::select! {
+			_ = self.inner.closed() => (),
+			_ = self.unsubscribe.unsubscribed() => (),
+		}
+	}
+
+	fn is_active_subscription(&self) -> bool {
+		!self.unsubscribe.is_unsubscribed()
+	}
+}
+
+impl Drop for SubscriptionSink {
+	fn drop(&mut self) {
+		if self.is_active_subscription() {
+			self.subscribers.lock().remove(&self.uniq_sub);
+		}
+	}
+}
+
+/// Wrapper struct that maintains a subscription "mainly" for testing.
+#[derive(Debug)]
+pub struct Subscription {
+	pub(crate) rx: mpsc::Receiver<String>,
+	pub(crate) sub_id: SubscriptionId<'static>,
+}
+
+impl Subscription {
+	/// Close the subscription channel.
+	pub fn close(&mut self) {
+		tracing::trace!("[Subscription::close] Notifying");
+		self.rx.close();
+	}
+
+	/// Get the subscription ID
+	pub fn subscription_id(&self) -> &SubscriptionId {
+		&self.sub_id
+	}
+
+	/// Returns `Some((val, sub_id))` for the next element of type T from the underlying stream,
```

Comment:
nit: Do we need to update the comment here? This returns Some(Ok((val, sub_id)) and does not panic anymore. Also, we could remove the `let res` below and return directly, I guess that was used for debugging :D

Reply:
no, it won't compile otherwise some lifetime issue without the let res binding ^^